Guard against racy channel allocation and connection close

domodwyer opened the pull request:
Proposes a solution to #251.

As the only write to `c.channels` is in `c.allocateChannel()` a simple `c.isClosed()` guard is sufficient to protect against concurrent writes - multiple readers is OK and does not require a lock.

Because `c.allocateChannel()` holds the connection mutex for the duration of the call and `c.shutdown()` acquires it after `c.closed` has been set, I don't believe there is any racy interaction using `c.closed`.

The test case is from the issue reported by @jmalloc - thanks it helped massively!

The changes to `channel.go` is a separate issue (and separate commit - I can remove it if you want) that eliminates a race between `channel.shutdown()` and `channel.call()`. It looks like the original code intended to avoid this race, but didn't get complete coverage - it became noticeable using `-race` after fixing the allocation race, so I included it in this PR.

Running `AMQP_URL="amqp://guest:guest@localhost:5672/" go test ./... -tags=integration -race` no longer produces any race warnings for me, though I still get:

```
--- FAIL: TestExchangeDeclarePrecondition (34.66s)
	integration_test.go:1759: dial integration server: Exception (501) Reason: "read tcp 127.0.0.1:58857->127.0.0.1:5672: i/o timeout"
```

I've always had this test case fail during integration tests and assume it's unrelated.

michaelklishin commented:
Thank you!

jmalloc commented:
FWIW, the tests that led me to create #251 originally are running smoothly using this PR.

streadway commented:
Looks good, but I'm concerned that this can deadlock when a server sent close arrives between the request and response of the call function. Can you add a test in `client_test.go` that would simulate the client send request, server recv request, server send close, server send response and make sure there isn't a deadlock?

michaelklishin commented:
JFYI, the easiest way of simulating a server-sent `connection.close` against a real RabbitMQ nodes is by closing a connection with `rabbitmqctl` or via management UI/HTTP API.

morus12 commented:
@michaelklishin 
I think that closing connection with `rabbitmqctl` or management is inconsistent and not testable.

Correct me if I'm wrong: 
a) a server close connection if not supported protocol header was received - non existing header can be added to force server to send close
b) rabbitmq close connection when publish is sent with mandatory flag to non existing exchange

jmalloc commented:
Looks like the usage of `ch.send` in `Ack()` is still racey when using the head of this PR :(
Please see [this CI build](https://travis-ci.org/rinq/rinq-go/jobs/219574702#L530), which is using 645418e5f0725c09ab98fde517e45865a369d352.

domodwyer commented:
Hi all, sorry for the delay, I've been out of the country for a few weeks.

Thanks @jmalloc, there were several calls to `channel.send()` that weren't covered by the mutex (`channek.Ack/Nack/Reject`) - I've added a test case to cover these racey calls.

Acquiring a lock for all Ack/Nack/Reject calls just to cover changing to `channel.sendClosed()` seems like overkill, so I've switched to using atomic operations to flip a flag controlling the behaviour of `channel.send()` instead of replacing the func pointer.

As the updated PR no longer aquires locks for switching the behaviour of channel.send(), does this resolve your concerns @streadway?

The test case I added failed with:
```
Dominics-MacBook-Pro:amqp % AMQP_URL="amqp://guest:guest@localhost:5672/" go test ./... -tags=integration -race -run TestRaceBetweenChannelShutdownAndSend
==================
WARNING: DATA RACE
Read at 0x00c4200682c8 by goroutine 16:
  github.com/domodwyer/amqp.(*Channel).Ack()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/channel.go:1505 +0x90
  github.com/domodwyer/amqp.TestRaceBetweenChannelShutdownAndSend.func1()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/connection_test.go:78 +0x46

Previous write at 0x00c4200682c8 by goroutine 13:
  github.com/domodwyer/amqp.(*Channel).shutdown.func1()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/channel.go:107 +0xe2
  sync.(*Once).Do()
      /usr/local/Cellar/go/1.8/libexec/src/sync/once.go:44 +0xe1
  github.com/domodwyer/amqp.(*Channel).shutdown()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/channel.go:138 +0xad
  github.com/domodwyer/amqp.(*Connection).closeChannel()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/connection.go:622 +0x42
  github.com/domodwyer/amqp.(*Channel).Close()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/channel.go:396 +0x19f

Goroutine 16 (running) created at:
  github.com/domodwyer/amqp.TestRaceBetweenChannelShutdownAndSend()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/connection_test.go:79 +0xe4
  testing.tRunner()
      /usr/local/Cellar/go/1.8/libexec/src/testing/testing.go:657 +0x107

Goroutine 13 (finished) created at:
  github.com/domodwyer/amqp.TestRaceBetweenChannelShutdownAndSend()
      /Users/dom/Documents/Go/src/github.com/domodwyer/amqp/connection_test.go:74 +0xb5
  testing.tRunner()
      /usr/local/Cellar/go/1.8/libexec/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestRaceBetweenChannelShutdownAndSend (0.01s)
	shared_test.go:68: channel close/send race close
	testing.go:610: race detected during execution of test
FAIL
FAIL	github.com/domodwyer/amqp	0.054s
```

and after commit 15912e66f832ac41d6b0bbc8fb436b16dc1f0f55:
```
Dominics-MacBook-Pro:amqp % AMQP_URL="amqp://guest:guest@localhost:5672/" go test ./... -tags=integration -race -run TestRaceBetweenChannelShutdownAndSend
ok  	github.com/domodwyer/amqp	1.060s
```

jmalloc commented:
No races detected from my tests so far :)

michaelklishin commented:
@jmalloc thank you for your feedback!

gerhard commented:
This definitely improves things as I can trigger race conditions easily without these changes. I'm still seeing a less rare race condition:

```
AMQP_URL="amqp://guest:guest@localhost:5672/" go test ./... -tags=integration -race -run=TestRaceBetweenChannelShutdownAndSend -count=10000
==================
WARNING: DATA RACE
Write at 0x00c420354f90 by goroutine 138:
  runtime.mapdelete()
      /usr/local/Cellar/go/1.8.1/libexec/src/runtime/hashmap.go:598 +0x0
  github.com/domodwyer/amqp.(*Connection).releaseChannel()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:600 +0xbc
  github.com/domodwyer/amqp.(*Connection).closeChannel()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:623 +0x70
  github.com/domodwyer/amqp.(*Channel).Close()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/channel.go:403 +0x19f

Previous read at 0x00c420354f90 by goroutine 159:
  runtime.mapiterinit()
      /usr/local/Cellar/go/1.8.1/libexec/src/runtime/hashmap.go:668 +0x0
  github.com/domodwyer/amqp.(*Connection).shutdown.func1()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:384 +0x23d
  sync.(*Once).Do()
      /usr/local/Cellar/go/1.8.1/libexec/src/sync/once.go:44 +0xe1
  github.com/domodwyer/amqp.(*Connection).shutdown()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:407 +0xcc
  github.com/domodwyer/amqp.(*Connection).dispatch0()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:431 +0x62c
  github.com/domodwyer/amqp.(*Connection).demux()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:414 +0x6e
  github.com/domodwyer/amqp.(*Connection).reader()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:508 +0x273

Goroutine 138 (running) created at:
  github.com/domodwyer/amqp.TestRaceBetweenChannelShutdownAndSend()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection_test.go:75 +0xb5
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107

Goroutine 159 (running) created at:
  github.com/domodwyer/amqp.Open()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:219 +0x578
  github.com/domodwyer/amqp.DialConfig()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:200 +0x414
  github.com/domodwyer/amqp.Dial()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection.go:132 +0x88
  github.com/domodwyer/amqp.integrationConnection()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/integration_test.go:1757 +0x50
  github.com/domodwyer/amqp.TestRaceBetweenChannelShutdownAndSend()
      /Users/gerhard/go/src/github.com/domodwyer/amqp/connection_test.go:70 +0x51
  testing.tRunner()
      /usr/local/Cellar/go/1.8.1/libexec/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestRaceBetweenChannelShutdownAndSend (0.01s)
	shared_test.go:68: channel close/send race close
	testing.go:610: race detected during execution of test
FAIL
FAIL	github.com/domodwyer/amqp	89.099s
```

The above might be out of scope for this PR, therefore I'm giving it 👍  since it definitely improves things.

Happy to merge @streadway?

domodwyer commented:
I've developed a fix for the rarer race @gerhard discovered (thanks!).

I'll open another PR after this one has been merged with a fix to keep everything clean, unless you want it in this PR too?

michaelklishin commented:
@domodwyer can you please rebase this against master? Thank you.

domodwyer commented:
Sorted!
